moveit_experimental/moveit_jog_arm/src/collision_check_thread.cpp

```diff
@@ -115,36 +124,72 @@ void CollisionCheckThread::startMainLoop(JogArmShared& shared_variables)
-            std::min(velocity_scale, exp(scene_velocity_scale_coefficient *
-                                         (scene_collision_distance - parameters_.scene_collision_proximity_threshold)));
+        // If we're definitely in collision, stop immediately
+        if (collision_detected)
+        {
+          velocity_scale = 0;
+        }
+
+        // If we are far from a collision, velocity_scale should be 1.
+        // If we are very close to a collision, velocity_scale should be ~zero.
+        // When scene_collision_proximity_threshold is breached, start decelerating exponentially.
+        if (scene_collision_distance < parameters_.scene_collision_proximity_threshold)
+        {
+          // velocity_scale = e ^ k * (collision_distance - threshold)
+          // k = - ln(0.001) / collision_proximity_threshold
+          // velocity_scale should equal one when collision_distance is at collision_proximity_threshold.
+          // velocity_scale should equal 0.001 when collision_distance is at zero.
+          velocity_scale = std::min(velocity_scale,
+                                    exp(scene_velocity_scale_coefficient *
+                                        (scene_collision_distance - parameters_.scene_collision_proximity_threshold)));
+        }
+
+        if (self_collision_distance < parameters_.self_collision_proximity_threshold)
+        {
+          velocity_scale =
+              std::min(velocity_scale, exp(self_velocity_scale_coefficient *
+                                           (self_collision_distance - parameters_.self_collision_proximity_threshold)));
+        }
       }
-
-      if (self_collision_distance < parameters_.self_collision_proximity_threshold)
+      // Else, we stop based on worst-case stopping distance
+      else
       {
-        velocity_scale =
-            std::min(velocity_scale, exp(self_velocity_scale_coefficient *
-                                         (self_collision_distance - parameters_.self_collision_proximity_threshold)));
+        // Retrieve the worst-case time to stop, based on current joint velocities
+
+        // Calculate rate of change of distance to nearest collision
+        current_collision_distance = std::min(scene_collision_distance, self_collision_distance);
+        derivative_of_collision_distance =
+            (current_collision_distance - prev_collision_distance) / collision_rate.expectedCycleTime().toSec();
+
+        if (current_collision_distance < parameters_.min_allowable_collision_distance &&
+            derivative_of_collision_distance <= 0)
+        {
+          velocity_scale = 0;
+        }
+        // Only bother doing calculations if we are moving toward the nearest collision
+        else if (derivative_of_collision_distance < 0)
+        {
+          // At the rate the collision distance is decreasing, how long until we collide?
+          est_time_to_collision = fabs(current_collision_distance / derivative_of_collision_distance);
```

Comment:
This should check for division by 0. Also, what happens if the distance values are not continues but a little bit noisy?

Reply:
I don't think there's a lot I can do about that. From my testing, the distance values have not been noisy, though.

I could apply a low-pass filter but the delay would not be helpful. I think it's best to leave this collision-checking mode optional for that reason. I can think of some scenarios where the user would not want dynamic collision checking (e.g. any contact task).

Reply:
I'll handle the zero-check within the `else if ()` above